Task/3 24 2026 json schema for filters documentation edition (#1682)
Alpine-DAV/ascent: src/docs/sphinx/developer_docs/Flow_Filter.rst

The Ascent runtime looks for the ``params`` node and passes it to the filter
upon creation. Parameters are verified when the filter is created during execution.

Filter Parameter Verification Schemas
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Filter parameters are validated and checked for any surprises using :ref:`param_schema`.

> In practice, a filter's interface ``param_schema`` describes the *shape* of the ``params``

Reviewer: ```suggestion
In practice, a filter's interface ``param_schema`` describes the expected structure of the ``params``
```

Author: Change applied in commit locally.

node that users provide when creating a filter.

.. note::
    In prior versions of Ascent, the ``verify_params`` method was used to allow the filter creator
    to define parameter verification and surprise checking for each individual flow filter.

>     This method now by default will reference the ``param_schema`` defined in the filter interface

Reviewer: ```suggestion
   The base implementation of ``verify_params`` now uses the ``param_schema`` defined in the filter interface
```

Author: Change applied in commit locally.

>     and execute both parameter validation and surprise checking against the defined schema

Reviewer: ```suggestion
    to validate parameters and check for surprises against the expected schema
```

Author: Change applied in commit locally.

    using :ref:`param_schema`.

Ascent Parameter Schema Helpers
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

To assist in constructing parameter schemas, Ascent has a number of  small schema
builder helpers defined in:

- ``src/libs/ascent/runtimes/flow_filters/ascent_runtime_param_check.hpp``
- ``src/libs/ascent/runtimes/flow_filters/ascent_runtime_param_check.cpp``

Each helper returns a Conduit node that represents a schema fragment (see :ref:`param_schema`).
These fragments are typically inserted under ``param_schema["properties/<param_name>"]``.

> 

Reviewer: if we change to ref style args, we will need to update these docs

Author: I updated these docs to reflect the updated access pattern.

        ps["required"].append() = "field";

        i["param_schema"].set(ps);
    }

string_schema
#############

Builds a string schema with optional length bounds.

.. code-block:: c++

    conduit::Node string_schema(optional_param<int> minLength = optional_param<int>(),
                                optional_param<int> maxLength = optional_param<int>());

.. raw:: html

  <details><summary><strong>Example: Using ``string_schema`` (with length bounds)</strong></summary>

.. code-block:: c++

    // Example: require a non-empty string with a max length
    conduit::Node ps;
    ps["type"] = "object";
    ps["additionalProperties"] = false;

    ps["properties/field"].set(string_schema(1, 64));
    ps["required"].append() = "field";

    i["param_schema"].set(ps);

Resulting schema fragment:

.. code-block:: yaml

  type: string
  minLength: 1
  maxLength: 64

Example input accepted by this schema:

.. code-block:: yaml

  params:
    field: "pressure"

.. raw:: html
  
  </details>
  <br>

string_enum_schema
##################

Builds a string schema restricted to an enumerated set of allowed values.

.. code-block:: c++

>     conduit::Node string_enum_schema(std::vector<std::string> options);

Reviewer: prefer const string ref vector

Author: Updated to be `conduit::Node &string_enum_schema(conduit::Node &schema_node, const std::vector<std::string> &options)`
